Implement a Multi-Lane Demo Using Hard-Coded Dragway

This will build upon #4989 . Perhaps using the "MOBIL" lane-change decision model.

This will also need a lane change planner. The planner could be as simple as specifying the time by which the car must be in the new lane.

Since #5032, which contains [EndlessRoadCar](https://github.com/maddog-tri/drake/blob/c8679ca8ac2b60cc25b779c3b48eb0eda1ef6add/drake/automotive/endless_road_car.h), has not been merged and I'm unable to compile the branch, I'm now looking to see whether I can simply plop a `SimpleCar` on top of a Dragway.

Here is the compile-time error I'm seeing when comiling @maddog-tri's "maliput-demo-carbits" branch:

```
[289/652] Building CXX object automotive/CMakeFiles/drakeAutomotive.dir/maliput/monolane/lane.cc.o
FAILED: /usr/bin/g++-4.9   -DEIGEN_MPL2_ONLY -DHAVE_SPDLOG -DdrakeAutomotive_EXPORTS -I/home/liang/dev/drake-distro-2/drake/.. -I/home/liang/dev/drake-distro-2/drake/pod-build/lcmgen -isystem /home/liang/dev/drake-distro-2/build/install/include/eigen3 -isystem /home/liang/dev/drake-distro-2/build/install/include -Werror=all -Werror=ignored-qualifiers -Werror=overloaded-virtual -Werror=extra -Werror=return-local-addr -Wno-unused-parameter -Wno-missing-field-initializers -g -fPIC   -std=gnu++14 -MMD -MT automotive/CMakeFiles/drakeAutomotive.dir/dev/endless_road_oracle.cc.o -MF automotive/CMakeFiles/drakeAutomotive.dir/dev/endless_road_oracle.cc.o.d -o automotive/CMakeFiles/drakeAutomotive.dir/dev/endless_road_oracle.cc.o -c /home/liang/dev/drake-distro-2/drake/automotive/dev/endless_road_oracle.cc
/home/liang/dev/drake-distro-2/drake/automotive/dev/endless_road_oracle.cc: In function ‘void drake::automotive::internal::IndexJunctions(const std::vector<drake::automotive::internal::SourceState>&, const std::vector<std::vector<drake::automotive::internal::PathRecord> >&, std::map<const drake::maliput::api::Junction*, std::vector<drake::automotive::internal::TimeBox> >*, std::map<int, std::vector<const drake::maliput::api::Junction*> >*)’:
/home/liang/dev/drake-distro-2/drake/automotive/dev/endless_road_oracle.cc:474:51: error: no matching function for call to ‘drake::automotive::internal::TimeBox::TimeBox(<brace-enclosed initializer list>)’
                                        s_in, s_out};
                                                   ^
/home/liang/dev/drake-distro-2/drake/automotive/dev/endless_road_oracle.cc:474:51: note: candidates are:
In file included from /home/liang/dev/drake-distro-2/drake/automotive/dev/endless_road_oracle.cc:2:0:
/home/liang/dev/drake-distro-2/drake/../drake/automotive/dev/endless_road_oracle-internal.h:62:8: note: drake::automotive::internal::TimeBox::TimeBox()
 struct TimeBox {
        ^
/home/liang/dev/drake-distro-2/drake/../drake/automotive/dev/endless_road_oracle-internal.h:62:8: note:   candidate expects 0 arguments, 6 provided
/home/liang/dev/drake-distro-2/drake/../drake/automotive/dev/endless_road_oracle-internal.h:62:8: note: constexpr drake::automotive::internal::TimeBox::TimeBox(const drake::automotive::internal::TimeBox&)
/home/liang/dev/drake-distro-2/drake/../drake/automotive/dev/endless_road_oracle-internal.h:62:8: note:   candidate expects 1 argument, 6 provided
/home/liang/dev/drake-distro-2/drake/../drake/automotive/dev/endless_road_oracle-internal.h:62:8: note: constexpr drake::automotive::internal::TimeBox::TimeBox(drake::automotive::internal::TimeBox&&)
/home/liang/dev/drake-distro-2/drake/../drake/automotive/dev/endless_road_oracle-internal.h:62:8: note:   candidate expects 1 argument, 6 provided
```

--- a/drake/automotive/BUILD
+++ b/drake/automotive/BUILD
@@ -180,6 +180,7 @@ drake_cc_library(
         ":generated_vectors",
         ":simple_car",
         ":trajectory_car",
+        "//drake/automotive/maliput/api",
         "//drake/automotive/maliput/utility",
         "//drake/lcm",
         "//drake/multibody/parsers",
@@ -203,6 +204,7 @@ drake_cc_binary(
     ],
     deps = [
         ":automotive_simulator",
+        "//drake/automotive/maliput/dragway",
         "//drake/common:text_logging_gflags",
     ],
 )
@@ -230,6 +232,7 @@ drake_cc_binary(
     deps = [
         ":automotive_simulator",
         ":infinite_circuit_road",
+        "//drake/automotive/maliput/dragway",
         "//drake/automotive/maliput/monolane:loader",
         "//drake/common:text_logging_gflags",
     ],

--- a/drake/automotive/CMakeLists.txt
+++ b/drake/automotive/CMakeLists.txt
@@ -103,7 +103,10 @@ if(lcm_FOUND)
       drake-rigid-body-plant)
 
   add_executable(automotive_demo automotive_demo.cc)
-  target_link_libraries(automotive_demo drakeAutomotiveLcm drakeLCMTypes)
+  target_link_libraries(automotive_demo
+      drakeAutomotiveLcm
+      drakeLCMTypes
+      drakeMaliputDragway)
 
   add_executable(car_sim_lcm car_sim_lcm.cc car_sim_lcm_common.cc)
   target_link_libraries(car_sim_lcm

--- a/drake/automotive/README.md
+++ b/drake/automotive/README.md
@@ -30,15 +30,16 @@ Running the demos
 -----------------
 
 A variety of demos are available.  In general, to run a demo, open a
-terminal and execute commands like this:
+terminal and execute commands like the ones shown below. End the demo by closing
+any window, or Ctrl-C in the terminal.  All of the launched programs will be
+closed.
 
 ```
 $ cd drake-distro
 $ bazel run drake/automotive:DEMO_NAME_HERE
 ```
 
-End the demo by closing any window, or Ctrl-C in the terminal.  All of the
-launched programs will be closed.
+The following demos are available:
 
  * Basic cars driving around on an open plane:
    `bazel run drake/automotive:demo`
@@ -63,6 +64,19 @@ launched programs will be closed.
 
    Same general idea as the `village-demo`, but with a different path for
    cars which includes a merge.
+
+ * A 10-lane dragway with a `TrajectoryCar` traveling down each lane plus one
+   `SimpleCar`:
+
+   ```
+   bazel run //drake/automotive:demo -- \
+       --num_dragway_lanes=10 \
+       --num_trajectory_car=10
+   ```
+
+   See the documentation for the parameters by running
+   `bazel run //drake/automotive:demo -- --help`.
+
 
 Driving the Prius
 -----------------

--- a/drake/automotive/automotive_demo.cc
+++ b/drake/automotive/automotive_demo.cc
@@ -6,6 +6,7 @@
 
 #include "drake/automotive/automotive_simulator.h"
 #include "drake/automotive/create_trajectory_params.h"
+#include "drake/automotive/maliput/dragway/road_geometry.h"
 #include "drake/common/drake_path.h"
 #include "drake/common/text_logging_gflags.h"
 
@@ -20,9 +21,23 @@ DEFINE_double(target_realtime_rate, 1.0,
 DEFINE_double(simulation_sec, std::numeric_limits<double>::infinity(),
               "Number of seconds to simulate.");
 
+DEFINE_int32(num_dragway_lanes, 0,
+             "The number of lanes on the dragway. The number of lanes is by "
+             "default zero to disable the dragway. A dragway road network is "
+             "only enabled when the user specifies a number of lanes greater "
+             "than zero.");
+DEFINE_double(dragway_length, 100, "The length of the dragway.");
+DEFINE_double(dragway_lane_width, 3.7, "The dragway lane width.");
+DEFINE_double(dragway_shoulder_width, 3.0, "The dragway's shoulder width.");
+
 namespace drake {
 namespace automotive {
 namespace {
+
+enum class RoadNetworkType {
+  flat = 0,
+  dragway = 1
+};
 
 std::string MakeChannelName(const std::string& name) {
   const std::string default_prefix{"DRIVING_COMMAND"};
@@ -32,10 +47,13 @@ std::string MakeChannelName(const std::string& name) {
   return default_prefix + "_" + name;
 }
 
-int main(int argc, char* argv[]) {
-  gflags::ParseCommandLineFlags(&argc, &argv, true);
-  logging::HandleSpdlogGflags();
-
+// Initializes the provided `simulator` with user-specified numbers of
+// `SimpleCar` vehicles and `TrajectoryCar` vehicles. If parameter
+// `road_network_type` equals `RoadNetworkType::dragway`, the provided
+// `road_geometry` parameter must not be `nullptr`.
+void AddVehicles(RoadNetworkType road_network_type,
+    const maliput::api::RoadGeometry* road_geometry,
+    AutomotiveSimulator<double>* simulator) {
   // TODO(liang.fok): Generalize this demo to allow arbitrary models to be
   // specified via command line parameters. This will involve removing some
   // hard-coded assumptions about the model's geometry. For exeample, the call
@@ -43,7 +61,6 @@ int main(int argc, char* argv[]) {
   // length and width.
   const std::string kSdfFile =
       GetDrakePath() + "/automotive/models/prius/prius_with_lidar.sdf";
-  auto simulator = std::make_unique<AutomotiveSimulator<double>>();
 
   if (FLAGS_simple_car_names.empty()) {
     const std::string name = "";
@@ -60,16 +77,97 @@ int main(int argc, char* argv[]) {
     }
   }
 
-  for (int i = 0; i < FLAGS_num_trajectory_car; ++i) {
-    const auto& params = CreateTrajectoryParams(i);
-    simulator->AddTrajectoryCarFromSdf(kSdfFile, std::get<0>(params),
-                                       std::get<1>(params),
-                                       std::get<2>(params));
+  if (road_network_type == RoadNetworkType::dragway) {
+    DRAKE_DEMAND(road_geometry != nullptr);
+    const maliput::dragway::RoadGeometry* dragway_road_geometry =
+        dynamic_cast<const maliput::dragway::RoadGeometry*>(road_geometry);
+    DRAKE_DEMAND(dragway_road_geometry != nullptr);
+    for (int i = 0; i < FLAGS_num_trajectory_car; ++i) {
+      const int lane_index = i % FLAGS_num_dragway_lanes;
+      const auto& params = CreateTrajectoryParamsForDragway(
+          *dragway_road_geometry, lane_index);
+      simulator->AddTrajectoryCarFromSdf(kSdfFile, std::get<0>(params),
+                                         std::get<1>(params),
+                                         std::get<2>(params));
+    }
+  } else {
+    for (int i = 0; i < FLAGS_num_trajectory_car; ++i) {
+      const auto& params = CreateTrajectoryParams(i);
+      simulator->AddTrajectoryCarFromSdf(kSdfFile, std::get<0>(params),
+                                         std::get<1>(params),
+                                         std::get<2>(params));
+    }
   }
+}
 
+// Adds a flat terrain to the provided `simulator`.
+void AddFlatTerrain(AutomotiveSimulator<double>* simulator) {
+  // Intentially do nothing. This is possible since only non-physics-based
+  // vehicles are supported and they will not fall through the "ground" when no
+  // flat terrain is present.
+  //
+  // TODO(liang.fok): Once physics-based vehicles are supported, a flat terrain
+  // should actually be added. This can be done by adding method
+  // `AutomotiveSimulator::AddFlatTerrain()`, which can then call
+  // `drake::multibody::AddFlatTerrainToWorld()`. This method is defined in
+  // drake/multibody/rigid_body_tree_construction.h.
+}
+
+// Adds a dragway to the provided `simulator`. The number of lanes, lane width,
+// lane length, and the shoulder width are all user-specifiable via command line
+// flags.
+const maliput::api::RoadGeometry* AddDragway(
+    AutomotiveSimulator<double>* simulator) {
+  std::unique_ptr<const maliput::api::RoadGeometry> road_geometry
+      = std::make_unique<const maliput::dragway::RoadGeometry>(
+          maliput::api::RoadGeometryId({"Automotive Demo Dragway"}),
+          FLAGS_num_dragway_lanes,
+          FLAGS_dragway_length,
+          FLAGS_dragway_lane_width,
+          FLAGS_dragway_shoulder_width);
+  return simulator->SetRoadGeometry(std::move(road_geometry));
+}
+
+// Adds a terrain to the simulated world. The type of terrain added depends on
+// the provided `road_network_type` parameter. A pointer to the road network is
+// returned. A return value of `nullptr` is possible if no road network is
+// added.
+const maliput::api::RoadGeometry* AddTerrain(RoadNetworkType road_network_type,
+    AutomotiveSimulator<double>* simulator) {
+  const maliput::api::RoadGeometry* road_geometry{nullptr};
+  switch (road_network_type) {
+    case RoadNetworkType::flat: {
+      AddFlatTerrain(simulator);
+      break;
+    }
+    case RoadNetworkType::dragway: {
+      road_geometry = AddDragway(simulator);
+      break;
+    }
+  }
+  return road_geometry;
+}
+
+// Determines and returns the road network type based on the command line
+// arguments.
+RoadNetworkType DetermineRoadNetworkType() {
+  if (FLAGS_num_dragway_lanes > 0) {
+    return RoadNetworkType::dragway;
+  } else {
+    return RoadNetworkType::flat;
+  }
+}
+
+int main(int argc, char* argv[]) {
+  gflags::ParseCommandLineFlags(&argc, &argv, true);
+  logging::HandleSpdlogGflags();
+  const RoadNetworkType road_network_type = DetermineRoadNetworkType();
+  auto simulator = std::make_unique<AutomotiveSimulator<double>>();
+  const maliput::api::RoadGeometry* road_geometry =
+      AddTerrain(road_network_type, simulator.get());
+  AddVehicles(road_network_type, road_geometry, simulator.get());
   simulator->Start(FLAGS_target_realtime_rate);
   simulator->StepBy(FLAGS_simulation_sec);
-
   return 0;
 }
 

--- a/drake/automotive/automotive_simulator.cc
+++ b/drake/automotive/automotive_simulator.cc
@@ -33,9 +33,11 @@
 #include "drake/systems/primitives/multiplexer.h"
 
 namespace drake {
-namespace automotive {
 
-using drake::multibody::joints::kRollPitchYaw;
+using multibody::joints::kFixed;
+using multibody::joints::kRollPitchYaw;
+
+namespace automotive {
 
 template <typename T>
 AutomotiveSimulator<T>::AutomotiveSimulator()
@@ -109,9 +111,8 @@ int AutomotiveSimulator<T>::AddTrajectoryCarFromSdf(
   builder_->Connect(*trajectory_car, *coord_transform);
   AddPublisher(*trajectory_car, vehicle_number);
   AddPublisher(*coord_transform, vehicle_number);
-  return AddSdfModel(sdf_filename, coord_transform, ""/*model_name*/);
+  return AddSdfModel(sdf_filename, coord_transform, ""/* model_name */);
 }
-
 
 template <typename T>
 int AutomotiveSimulator<T>::AddEndlessRoadCar(
@@ -162,6 +163,15 @@ int AutomotiveSimulator<T>::AddEndlessRoadCar(
 }
 
 template <typename T>
+const maliput::api::RoadGeometry* AutomotiveSimulator<T>::SetRoadGeometry(
+    std::unique_ptr<const maliput::api::RoadGeometry> road) {
+  DRAKE_DEMAND(!started_);
+  road_ = std::move(road);
+  GenerateAndLoadRoadNetworkUrdf();
+  return road_.get();
+}
+
+template <typename T>
 const maliput::utility::InfiniteCircuitRoad*
 AutomotiveSimulator<T>::SetRoadGeometry(
     std::unique_ptr<const maliput::api::RoadGeometry> road,
@@ -172,17 +182,21 @@ AutomotiveSimulator<T>::SetRoadGeometry(
   endless_road_ = std::make_unique<maliput::utility::InfiniteCircuitRoad>(
       maliput::api::RoadGeometryId({"ForeverRoad"}),
       road_.get(), start, path);
+  GenerateAndLoadRoadNetworkUrdf();
+  return endless_road_.get();
+}
 
+template <typename T>
+void AutomotiveSimulator<T>::GenerateAndLoadRoadNetworkUrdf() {
   maliput::utility::GenerateUrdfFile(road_.get(),
                                      "/tmp", road_->id().id,
                                      maliput::utility::ObjFeatures());
-  std::string urdf_filepath = std::string("/tmp/") + road_->id().id + ".urdf";
+  const std::string urdf_filepath =
+      std::string("/tmp/") + road_->id().id + ".urdf";
   parsers::urdf::AddModelInstanceFromUrdfFileToWorld(
       urdf_filepath,
       drake::multibody::joints::kFixed,
       rigid_body_tree_.get());
-
-  return endless_road_.get();
 }
 
 

--- a/drake/automotive/automotive_simulator.h
+++ b/drake/automotive/automotive_simulator.h
@@ -141,6 +141,12 @@ class AutomotiveSimulator {
       const maliput::api::LaneEnd& start,
       const std::vector<const maliput::api::Lane*>& path);
 
+  /// Sets the RoadGeometry for this simulation.
+  ///
+  /// @pre Start() has NOT been called.
+  const maliput::api::RoadGeometry* SetRoadGeometry(
+      std::unique_ptr<const maliput::api::RoadGeometry> road);
+
   /// Adds an LCM publisher for the given @p system.
   /// @pre Start() has NOT been called.
   void AddPublisher(const EndlessRoadCar<T>& system, int vehicle_number);
@@ -222,6 +228,11 @@ class AutomotiveSimulator {
   // model instances since each model instance has two entries: (1) its number
   // of position states and (2) its number of velocity states.
   std::vector<int> GetModelJointStateSizes() const;
+
+  // Generates the URDF model of the road network and loads it into the
+  // `RigidBodyTree`. Member variable `road_` must be set prior to calling this
+  // method.
+  void GenerateAndLoadRoadNetworkUrdf();
 
   // For both building and simulation.
   std::unique_ptr<RigidBodyTree<T>> rigid_body_tree_{

--- a/drake/automotive/create_trajectory_params.cc
+++ b/drake/automotive/create_trajectory_params.cc
@@ -3,6 +3,11 @@
 #include <algorithm>
 #include <vector>
 
+#include "drake/automotive/maliput/api/junction.h"
+#include "drake/automotive/maliput/api/lane.h"
+#include "drake/automotive/maliput/api/lane_data.h"
+#include "drake/automotive/maliput/api/segment.h"
+
 namespace drake {
 namespace automotive {
 
@@ -67,6 +72,27 @@ std::tuple<Curve2<double>, double, double> CreateTrajectoryParams(int index) {
   const double start_time = (index / curves.size()) * 0.8;
   const double kSpeed = 8.0;
   return std::make_tuple(curve, kSpeed, start_time);
+}
+
+std::tuple<Curve2<double>, double, double> CreateTrajectoryParamsForDragway(
+    const maliput::dragway::RoadGeometry& road_geometry, int index) {
+  const maliput::api::Segment* segment = road_geometry.junction(0)->segment(0);
+  DRAKE_DEMAND(index < segment->num_lanes());
+  const maliput::api::Lane* lane = segment->lane(index);
+  const maliput::api::GeoPosition start_geo_position =
+      lane->ToGeoPosition(maliput::api::LanePosition(
+          {0 /* s */, 0 /* r */, 0 /* h */}));
+  const maliput::api::GeoPosition end_geo_position =
+      lane->ToGeoPosition(maliput::api::LanePosition(
+          {lane->length() /* s */, 0 /* r */, 0 /* h */}));
+  std::vector<Curve2<double>::Point2> waypoints;
+  waypoints.push_back({start_geo_position.x, start_geo_position.y});
+  waypoints.push_back({end_geo_position.x, end_geo_position.y});
+  Curve2<double> curve(waypoints);
+
+  const double kSpeed = 8.0;
+  const double kStartTime = 0;
+  return std::make_tuple(curve, kSpeed, kStartTime);
 }
 
 }  // namespace automotive

--- a/drake/automotive/create_trajectory_params.h
+++ b/drake/automotive/create_trajectory_params.h
@@ -7,6 +7,8 @@
 #include <tuple>
 
 #include "drake/automotive/curve2.h"
+#include "drake/automotive/maliput/api/road_geometry.h"
+#include "drake/automotive/maliput/dragway/road_geometry.h"
 #include "drake/automotive/trajectory_car.h"
 
 namespace drake {
@@ -17,9 +19,23 @@ namespace automotive {
  * trajectory are not documented / promised by this API.
  *
  * @param index Selects which pre-programmed trajectory to use.
+ *
  * @return tuple of curve, speed, start_time
  */
 std::tuple<Curve2<double>, double, double> CreateTrajectoryParams(int index);
+
+/**
+ * Creates TrajectoryCar constructor demo arguments for a vehicle on a dragway.
+ * The details of the trajectory are not documented / promised by this API.
+ *
+ * @param road_geometry The dragway upon which the TrajectoryCar will travel.
+ *
+ * @param index The lane index within the provided `road_geometry`.
+ *
+ * @return tuple of curve, speed, start_time
+ */
+std::tuple<Curve2<double>, double, double> CreateTrajectoryParamsForDragway(
+    const maliput::dragway::RoadGeometry& road_geometry, int index);
 
 }  // namespace automotive
 }  // namespace drake

